### PR TITLE
Fix Binomial when p=0 and x=0

### PR DESCRIPTION
**Pull request by gBokiau**
`Binomial.dist(p=0.).logp(0)` should return 0. not -inf

### CODE REVIEW

**Comment by gBokiau** [comment]
Note: NUTS also throws 'bad initial energy: nan's' where `p = 1.` and `x < n`. 
I think that's a gradient issue, as logp = -inf in those cases. 

`p=1.` is bound to happen in logit/binomial regressions. Perhaps it's fair to say that a cloglog link might be better suited, as it is less prone to reaching the `1` limit.
In any case, a quick hack is `p=tt.clip(p, 0., 1.-1e-6)`.

Seeing this is fairly common, a more literate error message would come in handy and/or some alternate parametrisation that implements best practices for binomial regressions (much like `logit_p` does on Bernoulli dists). That being said, I'm not sure what that 'best practice' would be, and the literature I have at hand isn't very helpful in that regard — it tends to favour a logistic link function but doesn't mention that it is highly likely to generate unrealistic probabilities in MCMC inference.

 --
Follow up note  (to self, really)
- Not clear why NUTS does not generate nan's with `Bernoulli.dist(p=1).logp(0)` as it's essentially the same issue — will investigate. 
- Inverse Cauchy/arctan link is another good candidate to reduce the occurrence of values too close to 0 and 1 limits, albeit perhaps at more computational cost than cloglog. In practice and conceptually, a probability of 1 is more vexing  than one of 0 in the Binomial case: inferring that something is *entirely* impossible to happen sounds in many cases realistic, while an event that is *entirely* certain to happen isn't (this stands in favour of cloglog). However, the problem is symmetric in the Bernoulli case,  where there is little conceptual ground for an asymmetric error term.

**Comment by junpenglao** [comment]
I think this is a much need discussion. As you said many parametrization is great in 
MLE or MAP, but ill suit under MCMC

**Comment by gBokiau** [comment]
Looking at the Travis results, I take it `m` and `x` don't necessarily have the same shape. Not sure what the best approach is, should I force a broadcast?

File: pymc3/distributions/dist_math.py [hunk 1]
```diff
@@ -65,7 +65,7 @@ def logpow(x, m):
     Calculates log(x**m) since m*log(x) will fail when m, x = 0.
     """
     # return m * log(x)
-    return tt.switch(tt.eq(x, 0), -np.inf, m * tt.log(x))
+    return tt.switch(tt.eq(x, 0) & ~tt.eq(m, 0), -np.inf, m * tt.log(x))
```
Comment (twiecki) [inline pymc3/distributions/dist_math.py]: Does `tt.bitwise_and(tt.eq(x, 0), ~tt.eq(m, 0))` work better with the broadcast?

**Comment by twiecki** [comment]
> Seeing this is fairly common, a more literate error message would come in handy

Could we even detect that case? Perhaps in the likelihood itself.

**Comment by gBokiau** [comment]
Myeah, let's first see if NUTS can get past the nan's. Thanks for your review/suggestion,  `tt.and_()` and `tt.bitwise_and()` gave the same result, but I found a simple way to broadcast.

The [source](https://github.com/pymc-devs/pymc3/blob/master/pymc3/tests/test_examples.py#L270) for the test that used to fail says : 
> Fails on float32 due to starting inf at starting logP

Any idea why that would not be the case on float64?

**Comment by gBokiau** [comment]
NB this is otherwise good to merge for the p=0 and x=0 case.

**Comment by junpenglao** [comment]
Thanks @gBokiau !

**Comment by gBokiau** [comment]
So… it just struck me that `test_binomial` should not have previously worked, yet it did.
A cursory look at that source confirms the p=0, x=0 case is supposed to be tested.

Underperforming tests can scare me (especially for something this crucial), hence, while I don't have the time to further investigate this just this moment, I thought I'd at least bring this to your attention until I get the opportunity to have a deeper look.

**Comment by junpenglao** [comment]
Thanks @gBokiau, just had a quick look but p=0, x=0 is not tested:
```python
{'n': array(3), 'value': array(1), 'p': array(0.001)}
{'n': array(3), 'value': array(1), 'p': array(0.1)}
{'n': array(3), 'value': array(1), 'p': array(0.5)}
{'n': array(3), 'value': array(1), 'p': array(0.75)}
{'n': array(3), 'value': array(1), 'p': array(0.99)}
{'n': array(3), 'value': array(2), 'p': array(0.001)}
{'n': array(3), 'value': array(2), 'p': array(0.1)}
{'n': array(3), 'value': array(2), 'p': array(0.5)}
{'n': array(3), 'value': array(2), 'p': array(0.75)}
{'n': array(3), 'value': array(2), 'p': array(0.99)}
{'n': array(3), 'value': array(3), 'p': array(0.001)}
{'n': array(3), 'value': array(3), 'p': array(0.1)}
{'n': array(3), 'value': array(3), 'p': array(0.5)}
{'n': array(3), 'value': array(3), 'p': array(0.75)}
{'n': array(3), 'value': array(3), 'p': array(0.99)}
{'n': array(4), 'value': array(1), 'p': array(0.001)}
{'n': array(4), 'value': array(1), 'p': array(0.1)}
{'n': array(4), 'value': array(1), 'p': array(0.5)}
{'n': array(4), 'value': array(1), 'p': array(0.75)}
{'n': array(4), 'value': array(1), 'p': array(0.99)}
{'n': array(4), 'value': array(2), 'p': array(0.001)}
{'n': array(4), 'value': array(2), 'p': array(0.1)}
{'n': array(4), 'value': array(2), 'p': array(0.5)}
{'n': array(4), 'value': array(2), 'p': array(0.75)}
{'n': array(4), 'value': array(2), 'p': array(0.99)}
{'n': array(4), 'value': array(3), 'p': array(0.001)}
{'n': array(4), 'value': array(3), 'p': array(0.1)}
{'n': array(4), 'value': array(3), 'p': array(0.5)}
{'n': array(4), 'value': array(3), 'p': array(0.75)}
{'n': array(4), 'value': array(3), 'p': array(0.99)}
{'n': array(5), 'value': array(1), 'p': array(0.001)}
{'n': array(5), 'value': array(1), 'p': array(0.1)}
{'n': array(5), 'value': array(1), 'p': array(0.5)}
{'n': array(5), 'value': array(1), 'p': array(0.75)}
{'n': array(5), 'value': array(1), 'p': array(0.99)}
{'n': array(5), 'value': array(2), 'p': array(0.001)}
{'n': array(5), 'value': array(2), 'p': array(0.1)}
{'n': array(5), 'value': array(2), 'p': array(0.5)}
{'n': array(5), 'value': array(2), 'p': array(0.75)}
{'n': array(5), 'value': array(2), 'p': array(0.99)}
{'n': array(5), 'value': array(3), 'p': array(0.001)}
{'n': array(5), 'value': array(3), 'p': array(0.1)}
{'n': array(5), 'value': array(3), 'p': array(0.5)}
{'n': array(5), 'value': array(3), 'p': array(0.75)}
{'n': array(5), 'value': array(3), 'p': array(0.99)}
```

I put a `print(pt)` under `def check_logp(...)` in `class TestMatchesScipy(SeededTest):` to check this.

**Pull request merged**